src/robot/parsing/model/visitor.py

```diff
@@ -14,29 +14,63 @@
 #  limitations under the License.
 
 import ast
+from typing import Any, Callable, Iterator
 
+from ...utils.notset import NOT_SET, NotSet
 from .statements import Node
 
 
 class VisitorFinder:
+    __cls_finder_cache__: "dict[type[Any], Callable[..., Any]|None|NotSet]"
 
-    def _find_visitor(self, cls):
-        if cls is ast.AST:
+    def __new__(cls, *_args: Any, **_kwargs: Any):  # type: ignore[no-untyped-def]
+        if not hasattr(cls, "__cls_finder_cache__"):
+            cls.__cls_finder_cache__ = {}
+        return super().__new__(cls)
```

Comment:
1. Unless there's some concrete need to use `__new__`, this should be changed to `__init__`. It's less magical.
2. If `__new__` is needed, arguments it accepts should be passed forward to `super().__new__`.
3. `__dunder__` attributes shouldn't in general be used outside Python itself or specialized frameworks. We should just use `finder_cache`, `visitor_cache`, `visitor_method_cache`, or something like that. If you are afraid of attribute collision, using `__` prefix is fine.
4. `hasattr` could be avoided by initializing the cache to `None` on the class level. I guess it's a matter of taste does `not hasattr(foo, 'bar')` or `foo.bar is None` look better.

Reply:
Yeah, looks like a perfect use case for `__init_subclass__`. I had heard about it earlier, but have never used it for anything.

Reply:
ok, I changed it to __init_subclass__